### ISSUE TITLE
`dart pub publish` includes extraneous binary files

### ISSUE THREAD

**Issue by timsneath** [dart-lang/pub#3003]
I attempted to upload my package to the pub server, but it complained that it was too large after attempting to upload:

```
[C:\src\win32] dart pub publish
Publishing win32 2.1.1 to https://pub.dartlang.org:
|-- AUTHORS
|-- CHANGELOG.md
...
|-- example
|   |-- appcontainer.dart
|   |-- bin
|   |   |-- README.md
|   |   '-- taskdialog.exe.manifest
|   |-- bluetooth.dart
|   | (39 more...)
|   |-- window.dart
|   |-- winrt_picker.dart
|   '-- wmi.dart
...
|-- test
|   |-- api_test.dart
|   |-- btooth_test.dart
|   |-- com
|   |   |-- IApplicationActivationManager_test.dart
|   |   |-- IAppxFactory_test.dart
|   |   |-- IAppxFile_test.dart
|   |   | (67 more...)
|   |   |-- IWbemContext_test.dart
|   |   |-- IWbemLocator_test.dart
|   |   '-- IWbemServices_test.dart
|   | (17 more...)
|   |-- window_test.dart
|   |-- winrt_calendar_test.dart
|   '-- winrt_test.dart
'-- tool
    |-- README.md
...
    '-- struct_sizes.cpp

Publishing is forever; packages cannot be unpublished.
Policy details are available at https://pub.dev/policy

Do you want to publish win32 2.1.1 (y/N)? y
Uploading...
Uncompressed package archive is too large (size > 104857600).
```

On inspecting further, I discover that `dart pub publish` is attempting to upload various executables and symbol files that are artifacts of the build process. For example, 

- `build\test_cache\build\c075001b96339384a97db4862b8ab8db.cache.dill.track.dill`
- `example\explorer\windows\flutter\ephemeral\flutter_windows.dll.pdb` (170MB)

These are files in my `.gitignore`, and I'm surprised that they're being included. I am also surprised that they didn't cause a failure in the past, which makes me wonder if this might be a change of behavior in new builds...

```
[C:\src\win32] dart --version
Dart SDK version: 2.14.0-119.0.dev (dev) (Mon May 17 05:41:11 2021 -0700) on "windows_x64"
```

Full log: https://gist.github.com/timsneath/769172f3938c2a143c9f2d8366743820

**Comment by jonasfj** [comment]
Which Dart SDK version are you using?

We've rewritten gitignore handling... And server is better at enforcing size limits.

**Comment by sawankumarbundelkhandi** [comment]
very similar to https://github.com/dart-lang/pub/issues/2997. The dry run is not throwing any issue and is working fine but when I publishing I am getting the Uncompressed package archive is too large (size > 104857600) error.

**Comment by jonasfj** [comment]
This is separate from #2997, this likely has to do with `.gitignore` not being respected.

**Comment by mit-mit** [comment]
@timsneath can you please try with Dart 2.13.0 stable to confirm whether this is a 2.14 regression or not?

**Comment by jonasfj** [comment]
This should be a 2.14+ issue only, it [landed here with `.pubignore` support](https://github.com/dart-lang/sdk/commit/53448def7325b79e22f27438c7e57881ccf897a6), as this is when we started interpreting `.gitignore` ourselves.

Investigating this I found:
```
win32> git config --local --list --show-origin
file:.git/config        core.repositoryformatversion=0
file:.git/config        core.filemode=false
file:.git/config        core.bare=false
file:.git/config        core.logallrefupdates=true
file:.git/config        core.symlinks=false
file:.git/config        core.ignorecase=true
file:.git/config        remote.origin.url=https://github.com/timsneath/win32.git
file:.git/config        remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config        branch.main.remote=origin
file:.git/config        branch.main.merge=refs/heads/main
```

So at-least this repository in windows has `core.ignorecase=true`, 

Documentation for [`core.ignorecase´](https://git-scm.com/docs/git-config/2.14.6#Documentation/git-config.txt-coreignoreCase) says:
> The default is false, except git-clone[1] or git-init[1] will probe and set core.ignoreCase true if appropriate when the repository is created.

But the words: "if appropriate" are not qualified any further.

Maybe it's that `git clone` and `git init` on Windows will set `core.ignoreCase = true` in `.git/config` for the repository by default.
I'm thinking we could:
 * (A) Default to case-insensitive matching on Windows,
 * (B) Warn when we detect that git is configured with `core.ignoreCase = true`.

I'm leaning towards (A), if setting `core.ignoreCase = true` is indeed default on Windows. But how does Mac OS behave, is HFS also not case-insensitive?

**Comment by timsneath** [comment]
A is indeed true on an NTFS-formatted drive in Windows:

```
[C:\src\test2] git init
Initialized empty Git repository in C:/src/test2/.git/

[C:\src\test2] git config --local --list --show-origin
file:.git/config        core.repositoryformatversion=0
file:.git/config        core.filemode=false
file:.git/config        core.bare=false
file:.git/config        core.logallrefupdates=true
file:.git/config        core.symlinks=false
file:.git/config        core.ignorecase=true
```

**Comment by timsneath** [comment]
This is still occurring for my package. Has this been deployed?

**Comment by mit-mit** [comment]
> Has this been deployed?

I believe this is a client-side fix, so you'd need to upgrade to current master. Or did you already try that?

**Comment by timsneath** [comment]
Already on master, yes. It was trying to deploy the PDB and ephemeral
Windows directories when I uploaded an update to my package last night.

**Comment by mit-mit** [comment]
Looks like Dart build 2.14.0-165.0.dev [has a pub client](https://github.com/dart-lang/sdk/blob/2.14.0-165.0.dev/DEPS#L145) which [contains the commits](https://github.com/dart-lang/pub/commits/master) from May 21, that should have fixed it. Can you confirm you have 165 or later?

**Comment by timsneath** [comment]
[C:\src\win32] dart --version
Dart SDK version: 2.14.0-165.0.dev (dev) (Fri May 28 12:39:17 2021 -0700)
on "windows_x64"

**Comment by mit-mit** [comment]
OK, that sounds like an issue. Reopening, fyi @jonasfj

**Comment by jonasfj** [comment]
Did a bit of debugging on a windows instance, looks like we're not handling CR LF correctly in `.gitignore` files.

**Comment by jonasfj** [comment]
Fixed in #3026 rolls to SDK in [sdk/+/202481](https://dart-review.googlesource.com/c/sdk/+/202481).
